File: backend/app/schemas/reviews.py
"""Schemas for the shared evaluation review system."""
import uuid
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.base import CamelModel

ReviewDecision = Literal["accept", "reject", "correct"]
ReviewStatus = Literal["draft", "final"]


class ReviewEvidenceEntry(CamelModel):
    label: str
    value: str | list[str] | dict | None = None
    kind: Literal["text", "list", "json"] = "text"


class ReviewableAttribute(CamelModel):
    key: str
    label: str
    original_value: str | None = None
    allowed_values: list[str] = Field(default_factory=list)
    group: str | None = None
    source_label: str | None = None
    description: str | None = None
    evidence: str | None = None


class ReviewableItem(CamelModel):
    item_key: str
    item_type: str
    title: str
    subtitle: str | None = None
    badges: list[str] = Field(default_factory=list)
    evidence: list[ReviewEvidenceEntry] = Field(default_factory=list)
    attributes: list[ReviewableAttribute] = Field(default_factory=list)


class ReviewItemUpsert(CamelModel):
    item_key: str
    item_type: str
    attribute_key: str
    decision: ReviewDecision
    original_value: str | None = None
    reviewed_value: str | None = None
    reason_code: str | None = None
    note: str | None = None


class ReviewDraftUpdate(CamelModel):
    notes: str | None = None
    items: list[ReviewItemUpsert] = Field(default_factory=list)


class ReviewItemResponse(CamelModel):
    id: uuid.UUID
    item_key: str
    item_type: str
    attribute_key: str
    decision: ReviewDecision
    original_value: str | None = None
    reviewed_value: str | None = None
    reason_code: str | None = None
    note: str | None = None
    created_at: datetime
    updated_at: datetime


class ReviewSummaryResponse(CamelModel):
    id: uuid.UUID
    run_id: uuid.UUID
    reviewer_user_id: uuid.UUID
    reviewer_name: str | None = None
    status: ReviewStatus
    overall_decision: str | None = None
    notes: str | None = None
    review_snapshot: dict = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None


class ReviewDetailResponse(ReviewSummaryResponse):
    items: list[ReviewItemResponse] = Field(default_factory=list)


class RunReviewContextResponse(CamelModel):
    run_id: uuid.UUID
    app_id: str
    adapter: str
    item_types: list[str] = Field(default_factory=list)
    latest_review_id: uuid.UUID | None = None
    draft_review_id: uuid.UUID | None = None
    items: list[ReviewableItem] = Field(default_factory=list)
    history: list[ReviewSummaryResponse] = Field(default_factory=list)

File: src/types/reviews.ts
export type ReviewDecision = 'accept' | 'reject' | 'correct';
export type ReviewStatus = 'draft' | 'final';

export interface ReviewEvidenceEntry {
  label: string;
  value: string | string[] | Record<string, unknown> | null;
  kind: 'text' | 'list' | 'json';
}

export interface ReviewableAttribute {
  key: string;
  label: string;
  originalValue: string | null;
  allowedValues: string[];
  group?: string | null;
  sourceLabel?: string | null;
  description?: string | null;
  evidence?: string | null;
}

export interface ReviewableItem {
  itemKey: string;
  itemType: string;
  title: string;
  subtitle: string | null;
  badges: string[];
  evidence: ReviewEvidenceEntry[];
  attributes: ReviewableAttribute[];
}

export interface ReviewItemRecord {
  id: string;
  itemKey: string;
  itemType: string;
  attributeKey: string;
  decision: ReviewDecision;
  originalValue: string | null;
  reviewedValue: string | null;
  reasonCode: string | null;
  note: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface EvalReviewSummary {
  id: string;
  runId: string;
  reviewerUserId: string;
  reviewerName: string | null;
  status: ReviewStatus;
  overallDecision: string | null;
  notes: string | null;
  reviewSnapshot: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
}

export interface EvalReviewDetail extends EvalReviewSummary {
  items: ReviewItemRecord[];
}

export interface ActiveDraftInfo {
  reviewId: string;
  reviewerUserId: string;
  reviewerName: string | null;
  startedAt: string;
  isMine: boolean;
}

export interface RunReviewContext {
  runId: string;
  appId: string;
  adapter: string;
  itemTypes: string[];
  latestReviewId: string | null;
  draftReviewId: string | null;
  activeDraft: ActiveDraftInfo | null;
  items: ReviewableItem[];
  history: EvalReviewSummary[];
}

export interface ReviewItemUpsert {
  itemKey: string;
  itemType: string;
  attributeKey: string;
  decision: ReviewDecision;
  originalValue: string | null;
  reviewedValue: string | null;
  reasonCode: string | null;
  note: string | null;
}

export interface ReviewDraftUpdate {
  notes: string;
  items: ReviewItemUpsert[];
}
